### PR TITLE
[Publisher][Bug] Fix bug where adding team members to the admin panel clears out all other team members

### PR DESCRIPTION
**Pull request by mxosman**
## Description of the change

Fix bug where adding team members to the admin panel clears out all other team members due to recent optimizations of fetching team members separately.

This issue was raised by CSG members who noticed that adding a team member to an agency resulted in clearing out all other team members from the agency aside from the one recently added. Because we now fetch team members separately and ONLY when an agency card is clicked vs. before where they were populated for each agency immediately after the admin panel loads, we need to await the fetch of team members before updating them in the admin panel so all references to the team members will be the freshly fetched list.

Follow up TODO: Add unit tests to cover this area

## Type of change

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #XXXX

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx [hunk 1]
```diff
@@ -127,8 +127,15 @@ export const AgencyProvisioningOverview = observer(() => {
   const searchAndFilter = (e: React.ChangeEvent<HTMLInputElement>) => {
     setSearchInput(e.target.value);
   };
-  const editAgency = (agencyID: string | number) => {
-    const selectedAgency = agenciesByID[agencyID][0];
+  const editAgency = async (agencyID: string | number) => {
+    // Fetch the team associations for the agency.
+    await adminPanelStore.fetchAgencyTeam(String(agencyID));
```
Comment (mxosman) [inline publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx:132]: So, before when I said it was a one line change in the AgencyProvisioning, though it worked, I ended up going backt back further upstream to find the root of why things were out of sync since the AdminPanelStore eventually updates the agency's team member field that everything references. It turns out that we need to `await` the fetch populating the list of team members in the agency object, and use the `adminPanelStore.__` to stay in sync with the updated list we just fetched -- that way, by the time we include the team members in the store's `agencyProvisioningUpdates` in L149, it will be the fetched list and not an empty list.

LMK if this is confusing and I'll explain in a different way.

Comment (mxosman) [reply publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx:132]: This feels like a more proper solution though I hate that it waits for the fetch to open the modal. I'll think of a different solution or update a ticket I have for adding a loader to the Team Member & Roles tab to instead add a loader to the page while the modal opens.

**Review by lilidworkin — COMMENTED** [review]
This solution makes a lot of sense to me!

Did you test locally? Should we deploy to playtesting to test there to be safe, and also to see how bad the lag is?

Do we also need to make an equivalent change [here](https://github.com/Recidiviz/justice-counts/blob/23e9ccc9b1af3cd74847781d852dc554fa938313/publisher/src/components/AdminPanel/UserProvisioningOverview.tsx#L226-L227) for editing a user?

File: publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx (hunk 1 above)
Comment (lilidworkin) [reply publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx:132]: Yea this seems like a great solution! I think it's fine for now, as long as the lag isn't CRAZY bad. Maybe Ilya can add a loader a a fast-follow?

**Comment by mxosman** [comment]
> This solution makes a lot of sense to me!
> 
> Did you test locally? Should we deploy to playtesting to test there to be safe, and also to see how bad the lag is?

Yes, I tested locally + staging -- will deploy to playtesting for an extra sanity check!

> Do we also need to make an equivalent change [here](https://github.com/Recidiviz/justice-counts/blob/23e9ccc9b1af3cd74847781d852dc554fa938313/publisher/src/components/AdminPanel/UserProvisioningOverview.tsx#L226-L227) for editing a user?

Ah, good call! Let me take a look and see. I was scrambling yesterday afternoon and tunnel-visioned on this particular fix that I didn't think about the user provisioning.

**Comment by helperbot-recidiviz** [comment]
@mxosman successfully triggered a playtest deployment. Full deployment usually takes 5 minutes. Your playtest link is https://mahmoudtest---publisher-web-b47yvyxs3q-uc.a.run.app/

**Comment by mxosman** [comment]
> Do we also need to make an equivalent change [here](https://github.com/Recidiviz/justice-counts/blob/23e9ccc9b1af3cd74847781d852dc554fa938313/publisher/src/components/AdminPanel/UserProvisioningOverview.tsx#L226-L227) for editing a user?

Interestingly enough, this flow seems to work just fine. The core issue with the agency provisioning is here in L281:

https://github.com/Recidiviz/justice-counts/blob/23e9ccc9b1af3cd74847781d852dc554fa938313/publisher/src/components/AdminPanel/AgencyProvisioning.tsx#L280-L291

...where it's using the `agencyProvisioningUpdates.team` as a reference to the existing team members. The one line change that worked prior to this change was using `currentTeamMembers` instead of `agencyProvisioningUpdates.team`:

https://github.com/Recidiviz/justice-counts/blob/23e9ccc9b1af3cd74847781d852dc554fa938313/publisher/src/components/AdminPanel/AgencyProvisioning.tsx#L204-L211

I'll need to dig in more to better understand why the `agencyProvisioningUpdates` and the `selectedAgency` team are crossing the finish line at different times (and it seems that the `agencyProvisioningUpdates.team` doesn't ever quite get the update for some reason)...

The `UserProvisioning` only ever references the `selectedAgency` and not the `userProvisioningUpdates` for the current list of teams.

To keep everything in-line between the two flows, we could just go back to the original solution of using `currentTeamMembers` and everything will play nicely because it's referencing the `selectedAgency` and once the team is fetched and updated in the store, it will re-render with the proper team on `selectedAgency.team`. Whereas using the `agencyProvisioningUpdates.team` seems to not ever re-render with the proper team likely due to the store not updating the `agencyProvisioningUpdates.team` by the time it's fetched? I'm not 100% sure - would have to test that assumption. 

Either way, using the `currentTeamMembers` didn't quite feel like it was addressing the root cause at the time and I figured going a bit upstream would make more sense, but now that I'm writing this out and thinking more about it - it might make sense to stick with `currentTeamMembers` b/c the `selectedAgency` (or in the case of UserProvisioning - `selectedUser`) will always have the latest values.

Does any of my ramble make sense? Please let me know if it doesn't and I'll explain in a different way!

TLDR - I'm debating going back to the original solution and scrapping the changes in the overview so both UserProvisioning and AgencyProvisioning are congruent -- or updating the UserProvisioning to do the fetching in the same way here to keep that congruent. Both options will have the same results - it's just the latter blocks the modal opening until everything is fetched, whereas the former opens the modal and re-renders it when everything is fetched and updated in the store.

**Comment by lilidworkin** [comment]
@mxosman thank you for digging into this and for the super thoughtful explanation! I think I'll ultimately defer to you here. I'm tempted to say I like the cleanliness/safety of just blocking the modal opening until everything is fetched -- that feels like the most understandable solution and maybe the least prone to bugs getting introduced in the future? But I definitely don't feel strongly. +1 to being consistent across both flows, though.

**Comment by mxosman** [comment]
> @mxosman thank you for digging into this and for the super thoughtful explanation! I think I'll ultimately defer to you here. I'm tempted to say I like the cleanliness/safety of just blocking the modal opening until everything is fetched -- that feels like the most understandable solution and maybe the least prone to bugs getting introduced in the future? But I definitely don't feel strongly. +1 to being consistent across both flows, though.

OK - the more I think about it - it actually makes more sense to go with the one-line solution. That actually feels the most congruent across the board - both areas should reference the `selectedAgency`/`selectedUser` as the source of truth for the things that are being fetched separately so the UI re-renders when the store is updated. It doesn't make sense to make this same update in the UserProvisioning when that flow references the `selectedUser` and doesn't reference `userProvisioningUpdates.agencies` since `selectedUser` will always have and display the latest update. So the change in this PR accommodates the usage of `agencyProvisioningUpdates.team` in addition to the `selectedAgencies` (which is already naturally in sync). I don't think we need to also accommodate using a separate reference when we can (and imo should) use the central source of truth `selectedAgencies` for the additional fetch of team members.

Both modals technically wait for the fetch to complete - this additional `await` waits for the store to update before the modal is ready, which is not necessary b/c we can add a loader within the agencies box (for user provisioning) and in the team members tab (for agency provisioning) - there's already a task assigned to add the loaders in both areas.

Sorry for the back and forth here... I have the one-line solution PR here: https://github.com/Recidiviz/justice-counts/pull/1529 with a deploy to playtesting as well to sanity check. It works fine for me in staging.

**Comment by mxosman** [comment]
I'll record a loom that should hopefully explain this better than my words!

File: publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx (hunk 1 above)
Comment (morden35) [reply publisher/src/components/AdminPanel/AgencyProvisioningOverview.tsx:132]: this explanation/solution makes sense to me!

**Comment by mxosman** [comment]
https://www.loom.com/share/b657651b698b4430ac2ce3711b755b35 - OK I got cut off at 5 minutes but left a comment finishing the thought.

**Comment by mxosman** [comment]
Closing in favor of #1529
